arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -0,0 +1,299 @@
+    : ExecutorInfos(std::make_shared<std::unordered_set<RegisterId>>(readableInputRegisters),
+                    std::make_shared<std::unordered_set<RegisterId>>(writeableOutputRegisters),
+                    nrInputRegisters, nrOutputRegisters,
+                    std::move(registersToClear), std::move(registersToKeep)),
+      _aggregateVariables(aggregateVariables),
+      _aggregateRegisters(aggregateRegisters),
+      _groupRegisters(groupRegisters),
+      _collectRegister(collectRegister),
+      _groupVariables(std::move(groupVariables)),
+      _count(count),
+      _trxPtr(trxPtr) {
+  TRI_ASSERT(!_groupRegisters.empty());
+}
+
+std::vector<std::function<std::unique_ptr<Aggregator>(transaction::Methods*)> const*>
+HashedCollectExecutor::createAggregatorFactories(HashedCollectExecutor::Infos const& infos) {
+  std::vector<std::function<std::unique_ptr<Aggregator>(transaction::Methods*)> const*> aggregatorFactories;
+
+  if (infos.getAggregateVariables().empty()) {
+    // no aggregate registers. this means we'll only count the number of items
+    if (infos.getCount()) {
+      aggregatorFactories.emplace_back(
+          Aggregator::factoryFromTypeString("LENGTH"));
+    }
+  } else {
+    // we do have aggregate registers. create them as empty AqlValues
+    aggregatorFactories.reserve(infos.getAggregatedRegisters().size());
+
+    // initialize aggregators
+    for (auto const& r : infos.getAggregateVariables()) {
+      aggregatorFactories.emplace_back(
+          Aggregator::factoryFromTypeString(r.second.second));
+    }
+  }
+
+  return aggregatorFactories;
+}
+
+HashedCollectExecutor::HashedCollectExecutor(Fetcher& fetcher, Infos& infos)
+    : _infos(infos),
+      _fetcher(fetcher),
+      _upstreamState(ExecutionState::HASMORE),
+      _lastInitializedInputRow(InputAqlItemRow{CreateInvalidInputRowHint{}}),
+      _allGroups(1024,
+                 AqlValueGroupHash(_infos.getTransaction(),
+                                   _infos.getGroupVariables().size()),
+                 AqlValueGroupEqual(_infos.getTransaction())),
+      _producingOutput(false),
+      _isInitialized(false),
+      _aggregatorFactories() {
+  _aggregatorFactories = createAggregatorFactories(_infos);
+};
+
+HashedCollectExecutor::~HashedCollectExecutor() {
+  // Generally, _allGroups should be empty when the block is destroyed - except
+  // when an exception is thrown during getOrSkipSome, in which case the
+  // AqlValue ownership hasn't been transferred.
+  destroyAllGroupsAqlValues();
+}
+
+void HashedCollectExecutor::destroyAllGroupsAqlValues() {
+  for (auto& it : _allGroups) {
+    for (auto& it2 : it.first) {
+      const_cast<AqlValue*>(&it2)->destroy();
+    }
+  }
+}
+
+void HashedCollectExecutor::consumeInputRow(InputAqlItemRow& input) {
+  TRI_ASSERT(input.isInitialized());
+
+  decltype(_allGroups)::iterator currentGroupIt;
+  bool newGroup;
+  std::tie(currentGroupIt, newGroup) = findOrEmplaceGroup(input);
```

Comment:
`newGroup` is unused. I think it can be removed completely from the return value of `findOrEmplaceGroup`.